numpy/core/overrides.py

```diff
@@ -149,17 +153,18 @@ def verify_matching_signatures(implementation, dispatcher):
 def array_function_dispatch(dispatcher, module=None, verify=True):
     """Decorator for adding dispatch with the __array_function__ protocol."""
     def decorator(implementation):
-        # TODO: only do this check when the appropriate flag is enabled or for
-        # a dev install. We want this check for testing but don't want to
-        # slow down all numpy imports.
-        if verify:
-            verify_matching_signatures(implementation, dispatcher)
-
-        @functools.wraps(implementation)
-        def public_api(*args, **kwargs):
-            relevant_args = dispatcher(*args, **kwargs)
-            return array_function_implementation_or_override(
-                implementation, public_api, relevant_args, args, kwargs)
+        if not ENABLE_ARRAY_FUNCTION:
+            # __array_function__ requires an explicit opt-in for now
+            public_api = implementation
+        else:
+            if verify:
+                verify_matching_signatures(implementation, dispatcher)
+
+            @functools.wraps(implementation)
```

Comment:
Shouldn't the code that sets `__wrapped__` be inside this `else` too?

Reply:
Indeed, in fact, the setting of `__module__` is not needed either in the standard case.

Reply:
This was intentional: we need to set `__wrapped__` for the changes in `core.arrayprint` (probably I should add a comment).

That said, you're probably right here. The infinitely wrapped chain is problematic for `inspect.signature()`:
```
In [1]: import inspect

In [2]: def f(x):
   ...:     pass

In [3]: f.__wrapped__ = f

In [4]: inspect.signature(f)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-db9287230bbe> in <module>
----> 1 inspect.signature(f)

~/miniconda3/envs/numpy-py37/lib/python3.7/inspect.py in signature(obj, follow_wrapped)
   3068 def signature(obj, *, follow_wrapped=True):
   3069     """Get a signature object for the passed callable."""
-> 3070     return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
   3071
   3072

~/miniconda3/envs/numpy-py37/lib/python3.7/inspect.py in from_callable(cls, obj, follow_wrapped)
   2818         """Constructs Signature for the given callable object."""
   2819         return _signature_from_callable(obj, sigcls=cls,
-> 2820                                         follow_wrapper_chains=follow_wrapped)
   2821
   2822     @property

~/miniconda3/envs/numpy-py37/lib/python3.7/inspect.py in _signature_from_callable(obj, follow_wrapper_chains, skip_bound_arg, sigcls)
   2217     # Was this function wrapped by a decorator?
   2218     if follow_wrapper_chains:
-> 2219         obj = unwrap(obj, stop=(lambda f: hasattr(f, "__signature__")))
   2220         if isinstance(obj, types.MethodType):
   2221             # If the unwrapped object is a *method*, we might want to

~/miniconda3/envs/numpy-py37/lib/python3.7/inspect.py in unwrap(func, stop)
    513         id_func = id(func)
    514         if (id_func in memo) or (len(memo) >= recursion_limit):
--> 515             raise ValueError('wrapper loop when unwrapping {!r}'.format(f))
    516         memo[id_func] = func
    517     return func

ValueError: wrapper loop when unwrapping <function f at 0x10c08ff28>
```

(I will add a Python 3 only test for inspect.signature)

Reply:
> Indeed, in fact, the setting of `__module__` is not needed either in the standard case.

True, but I think this is a minor usability improvement regardless. I'm inclined to try to finish this up (see https://github.com/numpy/numpy/issues/12271) for NumPy 1.16 anyways.